.github/workflows/stage-3-build.yaml

```diff
@@ -33,37 +41,156 @@ on:
         type: string
 
 jobs:
-  artefact-1:
-    name: "Artefact 1"
+  build-and-push:
     runs-on: ubuntu-latest
-    timeout-minutes: 3
+    permissions:
+      id-token: write
+      contents: read
+    strategy:
+      matrix:
+        function: [process-pilot-data, notify]
     steps:
-      - name: "Checkout code"
-        uses: actions/checkout@v4
-      - name: "Build artefact 1"
+      - uses: actions/checkout@v4
+        with:
+          submodules: 'true'
+          fetch-depth: 0
+
+      - name: Az CLI login
+        if: github.ref == 'refs/heads/main'
+        uses: azure/login@v1
+        with:
+          client-id: ${{ secrets.AZURE_CLIENT_ID }}
+          tenant-id: ${{ secrets.AZURE_TENANT_ID }}
+          subscription-id: ${{ secrets.AZURE_SUBSCRIPTION_ID }}
+
+      - name: Azure Container Registry login
+        if: github.ref == 'refs/heads/main'
+        run: az acr login --name ${{ secrets.ACR_NAME }}
+
+      - name: Create Tags
+        env:
+          GH_TOKEN: ${{ github.token }}
+        continue-on-error: true
         run: |
-          echo "Building artefact 1 ..."
-      - name: "Check artefact 1"
+          echo "GITHUB_REF: $GITHUB_REF"
+          echo "The branch is: ${{ github.ref }}"
+
+          if [[ "$GITHUB_REF" == refs/pull/*/merge ]]; then
+            PR_NUM_TAG=$(echo "$GITHUB_REF" | sed 's/refs\/pull\/\([0-9]*\)\/merge/\1/')
+          else
+            PULLS_JSON=$(gh api /repos/{owner}/{repo}/commits/${GITHUB_SHA}/pulls)
+            ORIGINATING_BRANCH=$(echo ${PULLS_JSON} | jq -r '.[].head.ref' | python3 -c "import sys, urllib.parse; print(urllib.parse.quote_plus(sys.stdin.read().strip()))")
+            echo "ORIGINATING_BRANCH: $ORIGINATING_BRANCH"
+            PR_NUM_TAG=$(echo ${PULLS_JSON} | jq -r '.[].number')
+          fi
+
+          echo "PR_NUM_TAG: pr$PR_NUM_TAG"
+          echo "PR_NUM_TAG=pr$PR_NUM_TAG" >> $GITHUB_ENV
+
+          SHORT_COMMIT_HASH=$(git rev-parse --short $GITHUB_SHA)
+          echo "Commit hash tag: $SHORT_COMMIT_HASH"
+          echo "COMMIT_HASH_TAG=$SHORT_COMMIT_HASH" >> $GITHUB_ENV
+
+          echo "ENVIRONMENT_TAG=development" >> $GITHUB_ENV
+
+      - name: Build and Push Image
+        working-directory: ./
         run: |
-          echo "Checking artefact 1 ..."
-      - name: "Upload artefact 1"
+          function=${{ matrix.function }}
+
+          # Build the image
+          docker compose build $function
+
+          full_function_name="communication-management-$function"
+          repo_name="${{ secrets.ACR_NAME }}.azurecr.io/$full_function_name"
```

Comment:
@dnimmo @patrickmoore-nc We are missing the `ACR_NAME` secret and presumably some other GH secrets. 
See https://github.com/NHSDigital/dtos-communication-management/actions/runs/11579549346/job/32236087568